### PR TITLE
Missing promise features

### PR DESCRIPTION
**Pull request by brettz9**
Fixes part of issue #3. I want to make it easy for users of our promise-based db.js library to be able to keep working with a many features of promises as possible, requiring, in most cases, no other changes besides `var Promise = SyncPromise;`
- Breaking change (minor): Fix error messages upon sync resolved promise;
- Fix: Throw if `Promise.all` is resolved synchronously by all non-promises (and add test);
- Features: Add missing Promise features (2nd argument to then (making first arg. optional), Promise.race) and document/add tests;
- Features: Utilize `then(..., reject)` in place of `catch` for thenables like jQuery Deferred which don't support;
- Docs: Lint markdown;
- Docs: Show example where error is thrown when no catch;
- Docs: Fix README example for `all()` to use non-sync-resolving promise;
- Metadata: Remove deprecated "version" property from bower.json (should use tags);
- Metadata: Add contributors to `package.json` and `build.sh` to test script;
- Testing: Add mocha to `devDependencies` (and add `node_modules` to `.gitignore`);
- Testing: Add uncaught async exception test and one where no onRejected callback has been attached;
- Export: Use `self` where `this` is not available; use non-reserved `exprt` in place of `export`

### CODE REVIEW

**Comment by brettz9** [comment]
Not in any big rush, but just curious if you are maintaining the repo and might take a look at some point...

**Comment by paldepind** [comment]
Yes. I've already taken a quick look at it. The only thing I'm a bit u sure about is "Ensure Promise.all will not resolve synchronously". Why should promise all not resolve synchronously?

**Comment by brettz9** [comment]
As it had been previously, `SyncPromise.all([1, 2, 3])` would not only have a synchronous API but behave synchronously as well. Since `SyncPromise` itself returns a promise, I figured it should follow the same requirement of other SyncPromise's of behaving asynchronously (as per http://blog.ometer.com/2011/07/24/callbacks-synchronous-and-asynchronous/ ).

Btw, I rebased just now to remove one change I realized I no longer needed (the `setTimeout` in the `Promise.all` example) and to remove some timeouts I also no longer needed, so although it may look like `Promise.all` no longer has that change toward working asynchronously, other code (the line calling `SyncPromise.resolve()`) gives it that effect. I also added another test.

File: dist/sync-promise-amd.js [hunk 1]
```diff
@@ -26,58 +29,66 @@ function SyncPromise(fn) {
...
+      self.c[REJECTED].push(rej);
     }
   });
 };
 
 prot.catch = function(cb) {
   var self = this;
-  if (self.a) { // Promise has been resolved synchronously
-    throw new Error('Can not call catch on synchonously resolved promise');
-  }
```
Comment (paldepind) [inline dist/sync-promise-amd.js:80]: Why did you remove the error checking here? It's quite useful to ensure that a promise is not accidentally resolved synchronously.

File: README.md [hunk 2]
```diff
@@ -1,61 +1,57 @@
 # SyncPromise
 
-A fast, small, _safe_ promise implementation with synchronous promise resolution
-and an API which resembles ECMAScript promises.
+A fast, small, _safe_ promise implementation with synchronous promise
+resolution and an API which resembles `ECMAScript` promises.
```
Comment (paldepind) [inline README.md]: I'm not sure if code tags are appropriate in a README for things that aren't code.

**Comment by paldepind** [comment]
What it the use case for `SyncPromise.resolve` and `SyncPromise.reject`. They are not synchronous?

I've left a few inline comment. Besides the things mentioned I think this looks great.

File: dist/sync-promise-amd.js (hunk 1 above)
Comment (brettz9) [inline dist/sync-promise-amd.js:80]: Because I changed the behavior to force resolving asynchronously (and thus this code was no longer necessary).

File: README.md (hunk 2 above)
Comment (brettz9) [inline README.md]: I've seen it be used this way, but sure, I can remove, as yes, it is technically not code.

**Comment by brettz9** [comment]
They are as close to synchronous as possible without being synchronous (`setTimeout` of 0, though I suppose we could check first for `requestAnimationFrame` to see if it is available and use that if so). The use case is retaining some compatibility with existing promise scenarios or to say start with a blank promise to which an array of other promises (`then` or `catch`) can be attached in sequence.

Even this short duration seems it can be too long for a transaction though, so if you're open to breaking the rule about promises being consistently async or sync, I could make it truly sync.

**Comment by paldepind** [comment]
> They are as close to synchronous as possible without being synchronous (setTimeout of 0, though I suppose we could check first for requestAnimationFrame to see if it is available and use that if so).

That is no different from `Promise.resolve` and `Promise.reject`. Except these will actually fire before what you can achieve with `setTimeout` or `requestAnimationFrame`.

> Even this short duration seems it can be too long for a transaction though, so if you're open to breaking the rule about promises being consistently async or sync, I could make it truly sync.

Yes. The transaction will have closed. Breaking the rule about being consistently async is quite problematic. I don't think there is a proper way to implement `resolve` and `reject` which is why they currently are not.

**Comment by brettz9** [comment]
I've undone the backticks on non-code names. Give me a little while to undo the timeout changes.

If you have a _short_ reference or explanation demoing the problems with mixing sync and async, I'd appreciate it (I have "brain fog" from Chronic Fatigue Syndrome and from prior chemotherapy, so I have some gaps in sustaining concentration, as informative as the article you reference looks it would be).

**Comment by brettz9** [comment]
I've undone the timeout changes. Please check when you have a chance.

**Comment by brettz9** [comment]
And as far as the problems of mixing sync/async, I do see they could be problematic for debugging. 

I guess for those cases where I wish to immediately resolve or reject (e.g., if the user supplies bad arguments and we don't want to force the user to use try-catches), we might try making a useless `store.get` request or something like that which I suppose should keep the transaction alive...

**Comment by paldepind** [comment]
It looks great now. Thank you for the PR. I'll merge it in now.
